File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/js/JSAdditionalLine.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.js

import com.getcapacitor.JSObject
import com.microblink.core.AdditionalLine

/**
 * Represents a parsed additional line from a receipt. This class is part of the TIKI Inc.
 * SDK for receipt capture.
 *
 * @property type The type of the additional line.
 * @property text The text content of the additional line.
 * @property lineNumber The line number of the additional line within the receipt.
 */
class JSAdditionalLine(additionalLine: AdditionalLine) {
    private val type: JSStringType? = JSStringType.opt(additionalLine.type())
    private val text: JSStringType? = JSStringType.opt(additionalLine.text())
    private val lineNumber: Int = additionalLine.lineNumber()

    /**
     * Converts this `JSAdditionalLine` object to a JSON representation.
     *
     * @return A `JSONObject` representing the additional line.
     */
    fun toJS(): JSObject =
        JSObject()
            .put("type", type?.toJS())
            .put("text", text?.toJS())
            .put("lineNumber", lineNumber)

    companion object {
        /**
         * Creates an `JSAdditionalLine` object from an `AdditionalLine` if the input is not null.
         *
         * @param additionalLine The `AdditionalLine` to convert.
         * @return An `JSAdditionalLine` object, or null if the input is null.
         */
        fun opt(additionalLine: AdditionalLine?): JSAdditionalLine? =
            if (additionalLine != null) JSAdditionalLine(additionalLine) else null
    }
}

File: ios/Plugin/Plugin/Js/JSAdditionalLine.swift
/*
 * RspAdditionalLine Class
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import Capacitor
import BlinkReceipt
import BlinkEReceipt

/**
 Represents a response containing additional line information for a product in the ReceiptCapture plugin.

 This class is used to encapsulate additional line details of a product, including its type, text, and line number in response to plugin calls.
 */
class JSAdditionalLine {
    /// The type of the additional line (e.g., "Subtotal").
    private let type: JSStringType?
    
    /// The text content of the additional line.
    private let text: JSStringType?
    
    /// The line number representing the order of the additional line.
    private let lineNumber: Int

    /**
     Initializes an `RspAdditionalLine` instance based on a `BRProductAdditionalLine` object.

     - Parameter additionalLine: A `BRProductAdditionalLine` object containing additional line information.
     */
    init(additionalLine: BRProductAdditionalLine) {
        type = JSStringType(stringType: additionalLine.type)
        text = JSStringType(stringType: additionalLine.text)
        lineNumber = additionalLine.lineNumber
    }

    /**
     Converts the `RspAdditionalLine` object into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary representing the additional line's type, text, and line number in a format suitable for a Capacitor plugin call result.
     */
    func toJSObject() -> JSObject {
        var result = JSObject()
        result["type"] = type?.toJSObject()
        result["text"] = text?.toJSObject()
        result["lineNumber"] = lineNumber
        return result
    }
}
